packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts

```diff
@@ -0,0 +1,231 @@
+	chain: <T, V>(job: Job<T, X, V>, timeout?: number, importance?: JobImportance, id?: string) => IChainableJobQueue<T>
+	end: (onFulfilled?: FulfilledCallback<X>, onRejected?: RejectedCallback, onTimeout?: TimeoutCallback) => void
+}
+
+export interface IJobQueue {
+	get length(): number
+	clear(): void
+	endsWith(jobId: string): boolean
+	getLastJobIdByImportance(importance: JobImportance): string | undefined
+	enqueue<T, V>(
+		job: Job<T, void, V>,
+		timeout?: number,
+		importance?: JobImportance,
+		id?: string
+	): IChainableJobQueue<T>
+	removeJob(jobId: string): void
+}
+
+export class JobQueue extends EventEmitter implements IJobQueue {
+	private queue: JobDescription<any>[] = []
+	private currentJob: JobDescription<any> | null = null
+	private currentJobChain: JobDescription<any>[] = []
+	private currentJobTimeout: NodeJS.Timeout | null = null
+
+	constructor(private name: string, private logger: Logger) {
+		super()
+	}
+
+	get length(): number {
+		return this.queue.length
+	}
+
+	clear(): void {
+		if (this.currentJob) {
+			this.abortJob(this.currentJob)
+		} else {
+			if (this.currentJobChain.length) {
+				this.currentJob = this.currentJobChain[this.currentJobChain.length - 1]
+
+				Promise.allSettled(this.currentJobChain.map(async (job) => job.job.cleanup()))
+					.then(this.handleCleanupResults)
+					.finally(this.advanceJob)
+				this.removeChainedJobsFromTheFront()
+			}
+		}
+	}
+
+	enqueue<T, U>(
+		job: Job<T, void, U>,
+		timeout?: number,
+		importance?: JobImportance,
+		id?: string
+	): IChainableJobQueue<T> {
+		this.queue.push({ id, job, chained: false, timeout, importance, abortController: new AbortController() })
+		this.run()
+		return this as IChainableJobQueue<T>
+	}
+
+	chain<T, U, V>(
+		job: Job<T, U, V>,
+		timeout?: number,
+		importance?: JobImportance,
+		id?: string
+	): IChainableJobQueue<T> {
+		this.queue.push({ id, job, chained: true, timeout, importance, abortController: new AbortController() })
+		return this as IChainableJobQueue<T>
+	}
+
+	end<X>(onFulfilled?: FulfilledCallback<X>, onRejected?: RejectedCallback, onTimeout?: TimeoutCallback): void {
+		const lastJob = this.queue[this.queue.length - 1]
+		if (!lastJob) return
+		lastJob.onFulfilled = onFulfilled
+		lastJob.onRejected = onRejected
+		lastJob.onTimeout = onTimeout
+	}
+
+	getLastJobIdByImportance(importance: JobImportance): string | undefined {
+		for (let i = this.queue.length - 1; i >= 0; --i) {
+			if (this.queue[i].importance === importance) {
+				return this.queue[i].id
+			}
+		}
+		if (this.currentJob?.importance === importance) {
+			return this.currentJob.id
+		}
+		return undefined
+	}
+
+	removeJob(jobId: string): void {
+		this.queue = this.queue.filter((job) => job.id !== jobId && !job.chained)
+	}
+
+	endsWith(jobId: string): boolean {
+		const lastJob = this.queue[this.queue.length - 1] ?? this.currentJob
+		return lastJob?.id === jobId
+	}
+
+	private run(): void {
```

Comment:
Should `run` not make use of the importance, when choosing a job from the queue?

Reply:
That's why it's called _importance_ - it is not the same as priority, but If you have better ideas for naming, I'm all ears

Reply:
In this particular use case there should not be a scenario when a job with high importance is in the queue after a job with low importance. We just clear the queue before it can occur